### PR TITLE
cpu: Add clock source selection based on CLOCK_HSE or CLOCK_HSI for STM32F3 family

### PR DESCRIPTION
**Pull request by DipSwitch**
One more with #3703 in mind.

Propose to configure the STM32F3 PLL clock source with the CLOCK_HSE or CLOCK_HSI define in the includes\periph_conf.h from the BSP

Can't test HSI and HSE implementation, although nothing for HSE isn't changed.

For the HSI I figured that because you can only connect the HSI to the PLL with divide by 2 we must multiply the multiplier with 2 to get the requested clock speed.

### CODE REVIEW

**Comment by OlegHahm** [comment]
@haukepetersen, @PeterKietzmann, @thomaseichinger, anyone willing to review this?

**Comment by DipSwitch** [comment]
Tested and validated on the `stm32f3discovery`

File: cpu/stm32f3/cpu.c [hunk 1]
```diff
@@ -48,7 +68,7 @@ void cpu_init(void)
  */
 static void cpu_clock_init(void)
 {
-    /* configure the HSE clock */
+    /* configure the high speed clock */
```
Comment (PeterKietzmann) [inline cpu/stm32f3/cpu.c]: Do we need to set `HSION` in line 74?

File: cpu/stm32f3/cpu.c [hunk 2]
```diff
@@ -21,6 +22,25 @@
 #include "cpu.h"
 #include "periph_conf.h"
 
+/* Check the source to be used for the PLL */
+#if defined(CLOCK_HSI) && defined(CLOCK_HSE)
+#error "Only provide one of two CLOCK_HSI/CLOCK_HSE"
+#elif CLOCK_HSI
+#define CLOCK_CR_SOURCE             RCC_CR_HSION
+#define CLOCK_CR_SOURCE_RDY         RCC_CR_HSIRDY
+#define CLOCK_PLL_SOURCE            (RCC_CFGR_PLLSRC_HSI_DIV2)
```
Comment (PeterKietzmann) [inline cpu/stm32f3/cpu.c]: Naming scheme seems to be different for e.g. the nucleo-f303 board

Comment (PeterKietzmann) [inline cpu/stm32f3/cpu.c]: Compare [here](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32f3/include/stm32f303xe.h#L5574)

**Comment by PeterKietzmann** [comment]
For the protocol: Doesn't break things in the default case. Tested on a nudleo-f303 board

File: cpu/stm32f3/cpu.c (hunk 1 above)
Comment (DipSwitch) [inline cpu/stm32f3/cpu.c]: Nope guess not.

Comment (DipSwitch) [inline cpu/stm32f3/cpu.c]: Sinds this is only called from boot.

File: cpu/stm32f3/cpu.c (hunk 2 above)
Comment (DipSwitch) [inline cpu/stm32f3/cpu.c]: Addressed

**Comment by DipSwitch** [comment]
Tested both HSE 72MHz max and HSI 64 MHz max on stm32f3discovery.

File: cpu/stm32f3/cpu.c [hunk 3]
```diff
@@ -21,6 +22,30 @@
 #include "cpu.h"
 #include "periph_conf.h"
 
+/* Check the source to be used for the PLL */
+#if defined(CLOCK_HSI) && defined(CLOCK_HSE)
+#error "Only provide one of two CLOCK_HSI/CLOCK_HSE"
+#elif CLOCK_HSI
+#define CLOCK_CR_SOURCE             RCC_CR_HSION
+#define CLOCK_CR_SOURCE_RDY         RCC_CR_HSIRDY
+#ifdef RCC_CFGR_PLLSRC_HSI_DIV2
```
Comment (PeterKietzmann) [inline cpu/stm32f3/cpu.c]: Ugly :-P

**Comment by PeterKietzmann** [comment]
Works now. Let's wait for build tests..

File: cpu/stm32f3/cpu.c (hunk 3 above)
Comment (DipSwitch) [inline cpu/stm32f3/cpu.c]: Not much to do about it except to not automate it and make the user to through all the hussle :p

**Comment by OlegHahm** [comment]
Murdock is fine.

**Comment by PeterKietzmann** [comment]
ACK and go

**Pull request merged**